src/dataflow/src/sink/tail.rs

```diff
@@ -94,14 +98,68 @@ pub fn tail<G>(
                 return;
             }
         });
+
+        let progress_row = update_progress(
+            &mut input_frontier,
+            input.frontier().frontier(),
+            &mut packer,
+            connector.object_columns + 1,
+        );
+
+        // Only emit updates if this operator/worker received actual
+        // data for emission. For TAIL, data is exchanged to one worker,
+        // which forwards all the data to the client process. If we
+        // blindly forwarded the frontier from all workers we would get
+        // multiple progress updates in the client.
+        if connector.emit_progress && received_data && progress_row.is_some() {
+            let results = vec![progress_row.expect("known to exist")];
+
+            // TODO(benesch): the lack of backpressure here can result in
+            // unbounded memory usage.
+            if connector.tx.send(results).is_err() {
+                errored = true;
+                return;
+            }
+        }
     })
 }
 
-fn batch_upper(
-    batch: Option<&Rc<OrdValBatch<GlobalId, Row, u64, isize, usize>>>,
-) -> Option<Timestamp> {
-    batch
-        .map(|b| b.desc.upper().elements().get(0))
-        .flatten()
-        .copied()
+// Checks if there is progress between `current_input_frontier` and
+// `new_input_frontier`. If yes, updates the `current_input_frontier` to
+// `new_input_frontier` and returns a [`Row`] that encodes this progress. If
+// there is no progress, returns [`None`].
+fn update_progress(
+    current_input_frontier: &mut Vec<Timestamp>,
+    new_input_frontier: AntichainRef<Timestamp>,
+    packer: &mut Row,
+    empty_columns: usize,
+) -> Option<Row> {
+    // Test to see if strict progress has occurred (any of the old frontier less equal
+    // to the new frontier).
+    let progress = current_input_frontier
```

Comment:
timely provides a `less_than` implementation for antichain with `current_input_frontier.frontier().less_than(&new_input_frontier.frontier())`

Reply:
I took this one straight from DD: https://github.com/timelydataflow/differential-dataflow/blob/87ed7e7ab2428b9b05f1aba90c5ddb2def582ac0/src/operators/arrange/arrangement.rs#L596

Reply:
I think the reason is that `Antichain.extend()` is slightly more complicated than just extending a `Vec`.

Reply:
it looks like an omission there too, I pushed a PR https://github.com/TimelyDataflow/differential-dataflow/pull/328

Reply:
Thanks! I'll update. :+1:

Reply:
Yeah sorry. The `PartialOrd` implementation showed up later than the arrange implementation, and while various things were pivoted over (e.g. the various trace handles) this was missed!